modules/metastorage/src/main/java/org/apache/ignite/internal/metastorage/impl/MetaStorageRaftGroupEventsListener.java

```diff
@@ -0,0 +1,247 @@
+     *
+     * <p>Multi-threaded access is guarded by {@code serializationFutureMux}.
+     */
+    private CompletableFuture<Void> serializationFuture = null;
+
+    private final Object serializationFutureMux = new Object();
+
+    /**
+     * The most recent version of the Logical Topology that this listener has observed.
+     */
+    private volatile long lastProcessedLogicalTopologyVersion = -1;
+
+    MetaStorageRaftGroupEventsListener(
+            IgniteSpinBusyLock busyLock,
+            ClusterService clusterService,
+            LogicalTopology logicalTopology,
+            CompletableFuture<MetaStorageServiceImpl> metaStorageSvcFut
+    ) {
+        this.busyLock = busyLock;
+        this.topologyService = clusterService.topologyService();
+        this.logicalTopology = logicalTopology;
+        this.metaStorageSvcFut = metaStorageSvcFut;
+    }
+
+    @Override
+    public void onLeaderElected(long term) {
+        registerTopologyEventListeners();
+
+        // Update learner configuration in case we missed some topology updates.
+        executeIfLeader(service -> removeLearners(service.raftGroupService(), logicalTopology.getLogicalTopology()));
+    }
+
+    private void registerTopologyEventListeners() {
+        // Try to remove a Meta Storage Learner if a node leaves the Physical Topology. This handles the case when a node joins, but leaves
+        // before sending the Join Ready request (i.e. before it gets added to the Logical Topology).
+        topologyService.addEventHandler(new TopologyEventHandler() {
+            @Override
+            public void onDisappeared(ClusterNode member) {
+                executeIfLeader(service -> removeLearners(service.raftGroupService(), logicalTopology.getLogicalTopology()));
+            }
+        });
+
+        logicalTopology.addEventListener(new LogicalTopologyEventListener() {
+            // Try to add a node as a Learner when it appears in the Logical Topology. Even though usually remote nodes add themselves as
+            // learners, this call allows us to handle the case when a network partition occurs and a node gets kicked out of the Logical
+            // Topology. Since the remote node simply gets disconnected and not restarted, it should be able to re-join the cluster even
+            // though it will skip the step when it adds itself as a Learner.
+            @Override
+            public void onAppeared(ClusterNode appearedNode, LogicalTopologySnapshot newTopology) {
+                executeIfLeader(service -> addLearner(service.raftGroupService(), appearedNode));
+            }
+
+            // Try to remove a Meta Storage Learner when a node leaves the Logical Topology. This approach protects from network glitches
+            // when a node leaves and re-joins the Physical Topology.
+            @Override
+            public void onDisappeared(ClusterNode disappearedNode, LogicalTopologySnapshot newTopology) {
```

Comment:
Seems that we don't have happens before guarantees here. In other words it's possible to handle logical.onDisappeared(A) while processing physical.onAppeared(A) just because one is about in-raft CMG thread and another is network topology internals.

Reply:
Again, can you please clarify the situation a little bit?

Reply:
My fault. We do noting on physical.onAppeared(A), so there's no race.

Reply:
However, seems, that we may have a race on CMG leadership transfer.
1. C joins the topology and register itself as a learner.
2. B becomes MS leader and fires executeIfLeader(service -> removeLearners(service.raftGroupService(), logicalTopology.getLogicalTopology())); However B doesn't know yet that C is part of logical topology, so it removes it.
3. Yep, you do expect that when onAppeared event will be fired you'll add C (which is wrong from my point of view), however...
4. ...A becomes MS leader, without B receiving  onAppeared(C), and A already received onAppeared(C), so no one will restore C as a learner.